ipld/read.go

```diff
@@ -18,190 +17,170 @@ import (
-			rootCid, err := plugin.CidFromNamespacedSha256(rowRoots[row])
-			if err != nil {
-				return nil, err
-			}
+	dataSquare := make([][]byte, edsWidth*edsWidth)
 
-			go sc.retrieveShare(rootCid, true, row, col, dag)
-		}
-	}
-	// wait until enough data has been collected, too many errors encountered,
-	// or the timeout is reached
-	err := sc.wait()
+	quadrant, err := pickRandomQuadrant(rowRoots)
 	if err != nil {
 		return nil, err
 	}
-	// flatten the square
-	flattened := sc.flatten()
-	// repair the square
-	tree := wrapper.NewErasuredNamespacedMerkleTree(uint64(edsWidth) / 2)
-	return rsmt2d.RepairExtendedDataSquare(rowRoots, colRoots, flattened, codec, tree.Constructor)
-}
-
-// uniqueRandNumbers generates count unique random numbers with a max of max
-func uniqueRandNumbers(count, max int) []uint32 {
-	if count > max {
-		panic(fmt.Sprintf("cannot create %d unique samples from a max of %d", count, max))
-	}
-	samples := make(map[uint32]struct{}, count)
-	for i := 0; i < count; {
-		// nolint:gosec // G404: Use of weak random number generator
-		sample := uint32(rand.Intn(max))
-		if _, has := samples[sample]; has {
-			continue
-		}
-		samples[sample] = struct{}{}
-		i++
+	errGroup, ctx := errgroup.WithContext(parentCtx)
+	errGroup.Go(func() error {
+		return fillQuadrant(ctx, quadrant, dag, dataSquare)
+	})
+	if err := errGroup.Wait(); err != nil {
+		return nil, err
 	}
-	out := make([]uint32, count)
-	counter := 0
-	for s := range samples {
-		out[counter] = s
-		counter++
+	batchAdder := NewNmtNodeAdder(parentCtx, ipld.NewBatch(parentCtx, dag, ipld.MaxSizeBatchOption(batchSize(edsWidth))))
+	tree := wrapper.NewErasuredNamespacedMerkleTree(uint64(edsWidth)/2, nmt.NodeVisitor(batchAdder.Visit))
+	extended, err := rsmt2d.RepairExtendedDataSquare(dah.RowsRoots, dah.ColumnRoots, dataSquare, codec, tree.Constructor)
+	if err != nil {
+		return nil, err
 	}
-	return out
-}
 
-type index struct {
-	row uint32
-	col uint32
+	return extended, batchAdder.Commit()
 }
 
-type indexedShare struct {
-	data []byte
-	index
+type quadrant struct {
+	isLeftSubtree bool
+	from          int
+	rootCids      []cid.Cid
 }
 
-// shareCounter is a thread safe tallying mechanism for share retrieval
-type shareCounter struct {
-	// all shares
-	shares map[index][]byte
-	// number of shares successfully collected
-	counter uint32
-	// the width of the extended data square
-	edsWidth uint32
-	// the minimum shares needed to repair the extended data square
-	minSharesNeeded uint32
+func pickRandomQuadrant(roots [][]byte) (*quadrant, error) {
+	edsWidth := len(roots)
+	// get the random number from 1 to 4.
+	q := rand.Intn(4) + 1 //nolint:gosec
+	quadrant := &quadrant{rootCids: make([]cid.Cid, edsWidth/2)}
+	// quadrants 1 and 3 corresponds to left subtree,
+	// 2 and 4 to the right subtree
+	/* | 1 | 2 |
+	   | 3 | 4 |
+	*/
+	// choose subtree
+	if q%2 == 1 {
+		quadrant.isLeftSubtree = true
+	}
+	to := 0
```

Comment:
I don't understand why `from` is a field inside of `quadrant` but `to` is not, and vice versa. Is there a reason?

Reply:
We need `from` field in fillQuadrant to correctly set index of share. But `to`  var is used to define the last root that we need to store from RowRoots, since the quadrant can have two different ranges:
* [0;edsWidth/2)
* [edsWidth/2;edsWidth)

Reply:
I still don't understand why `to` can't also be a field set within `quadrant` ? it's just a bit hard to follow the code here.

Reply:
extended `quadrant` struct with `to` field

Reply:
But neither does `quadrant.from` right?

Reply:
It is  used to calculate `position` in `dataSquare`